plugins/builds/index.js

```diff
@@ -870,19 +1058,77 @@ const buildsPlugin = {
+                        if (nextBuildStageName) {
+                            const nextBuildStage = await stageFactory.get({
+                                pipelineId: current.pipeline.id,
+                                name: nextBuildStageName
+                            });
+                            const stageSetupName = getFullStageJobName({
+                                stageName: nextBuildStageName,
+                                jobName: 'setup'
+                            });
+
+                            // If next build is stage setup, create stageBuild and set to running
+                            if (nextBuildStage && nextJobName === stageSetupName) {
+                                await stageBuildFactory.create({
+                                    stageId: nextBuildStage.id,
+                                    eventId: current.event.id,
+                                    status: 'CREATED'
+                                });
+                            }
+                        }
+
                         return createInternalBuild(internalBuildConfig);
                     }
 
-                    if (!['CREATED', null, undefined].includes(existNextBuild.status)) {
+                    if (!['CREATED', null, undefined].includes(existNextBuild.status) && !current.stage) {
                         return existNextBuild;
                     }
 
+                    // Current build is part of stage
+                    if (current.stage) {
+                        const stageBuild = await stageBuildFactory.get({
+                            stageId: current.stage.id,
+                            eventId: current.event.id
+                        });
+                        const stageTeardownName = getFullStageJobName({
+                            stageName: current.stage.name,
+                            jobName: 'teardown'
+                        });
+
+                        //   if currentBuild is stage teardown with success, check stageBuild status
+                        if (current.job.name === stageTeardownName && current.build.status === 'SUCCESS') {
+                            //      if stageBuild status is failure/terminal,
+                            if (['FAILURE', 'ABORTED', 'COLLAPSED', 'UNSTABLE'].includes(stageBuild.status)) {
+                                // Remove next build
+                                logger.info(
+                                    `Failure occurred in upstream job, removing new build - build:${existNextBuild.id} pipeline:${current.pipeline.id}-${current.job.name} event:${existNextBuild.eventId} `
+                                );
+                                await existNextBuild.remove();
+
+                                return null;
+                            }
+
+                            return existNextBuild;
+                        }
+                    }
+
+                    // Current build is not part of stage
```

Comment:
With the latest changes in `index.js` update build endpoint, `triggerNextJobs()` will not be invoked when `teardown` has successfully completed, but `stageBuild` has failure. Hence, this block of code will not clean up the downstream builds of `teardown`.

Instead, invoke `removeJoinBuilds()` in `index.js` update build endpoint to cleanup downstream builds of `teardown` when `stageBuild` has failure.
```suggestion
                    if (!['CREATED', null, undefined].includes(existNextBuild.status)) {
                        return existNextBuild;
                    }
                    
```